### ISSUE TITLE
"loginctl lock-session" fails when no explicit session passed

### ISSUE THREAD

**Issue by WhyNotHugo** [systemd/systemd#6032]
### Submission type

 - Bug report

### systemd version the issue has been seen with

```
$ systemctl --version
systemd 233
+PAM -AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN default-hierarchy=hybrid

```

### Used distribution

```
$ uname -a
Linux athena.barrera.io 4.11.2-1-ARCH #1 SMP PREEMPT Mon May 22 06:53:49 CEST 2017 x86_64 GNU/Linux
```

### Bug details

`lock-session` doesn't lock the current session if no parameters are provided (just a very unclear error), though the docs seem to specify that the current session is used if no arguments are provided:

```
$ loginctl lock-session
Failed to issue method call: No such device or address
```

Manually specifying the session does works (well, it doesn't do any locking, but it does fire the signal):

```
$ loginctl lock-session c1
$
```

The session is my current (and only) local session.

```
$ loginctl list-sessions 
   SESSION        UID USER             SEAT             TTY             
        c1       1000 hugo             seat0            tty1            

1 sessions listed.
$
```

Everything else session-related runs okay, and the `systemd --user` instance is also running (so `pam_systemd` wouldn't appear to be failing to start).

**Comment by poettering** [comment]
Hmm, invoking "loginctl lock-sessions" should indeed lock the session of the caller. Something appears broken there.

**Comment by WhyNotHugo** [comment]
`sudo loginctl lock-sessions` (eg: plural "sessions") does indeed fire the signal for my current session. It's the singular form that raises the unclear error (rather than lock the current session).

**Comment by mbiebl** [comment]
Is the ```XDG_SESSION_ID``` env variable properly set in the shell where you run ```loginctl lock-session```.
My guess is, that you try to (wrongly) run lock-session via sudo, where XDG_SESSION_ID is unset.

**Comment by WhyNotHugo** [comment]
Nope, I'm not sudoing `loginctl lock-session`. Only `lock-sessions`. I *though* `XDG_SESSION_ID` was unset, but apparently it is:

```
$ echo $XDG_SESSION_ID 
c1
$ loginctl lock-session
Failed to issue method call: No such device or address
```

**Comment by keszybz** [comment]
`loginctl lock-session` does not use `$XDG_SESSION_ID` in any way. It just calls `org.freedesktop.login1.Manager.LockSession` with "" as the session argument, and `logind` tries to find out the session. If it cannot determine the session of the calling program, you get the error you got.

Are you calling `lock-session` from `gnome-terminal`? `gnome-terminal` runs as a daemon in the `systemd --user` service unit, so it's not part of a session. Does this work if you call it from an `xterm` started from alt-f2?

I see the following solutions:
1. just say that this is working as expected
2. try to improve the error message, e.g. translate `ENXIO` to "Cannot determine the session of the caller"
3. make loginctl smarter, and use `$XDG_SESSION_ID` if set
4. do 3 and 4 together.

**Comment by WhyNotHugo** [comment]
> Are you calling lock-session from gnome-terminal? gnome-terminal runs as a daemon in the systemd --user service unit, so it's not part of a session. Does this work if you call it from an xterm started from alt-f2?

Nope, I'm not running gnome-terminal (I'm using termite). **However**, since my window manager and runners (eg: rofi, sxhkd) run as a `systemd --user` services, any apps I spawn are actually spawned inside that service's cgroup/context, so I'm guessing that I've a similar scenario.

Considering I run my WM and launcher (sxhkd) as systemd services, how would I make child processes part of my session so locking works via them? I honestly don't know how else to spawn processes.

**Comment by WhyNotHugo** [comment]
Oh, the error message definitely needs improvement, regardless of whether (3) is acceptable or not.

**Comment by WhyNotHugo** [comment]
`$XDG_SESSION_ID` isn't present in my environment so the related PR wouldn't change anything for. It seems that `pam_systemd` simply isn't setting the variable. Any hints?

**Comment by keszybz** [comment]
`pam_systemd` most likely sets it, but either you're running in an environment that is not a child of the login process, or something unset it.

**Comment by WhyNotHugo** [comment]
Yeah, so I just tested this, and here's the problem:

* When my X session starts, I run `sxhkd` (a "simple hotkey daemon") via `systemd --user`.
* I then press <kbd>super</kbd>+<kbd>t</kbd>, and `sxhkd` opens a terminal.
  * This terminal is now a child of `sxhkd`, which is systemd-managed, and not a child of the login process.
* This terminal inherits none of the login env variables.

There's a twist to all this though:
* I run `systemctl --user import-environment` **before** starting the `sxhkd` unit file, but it looks like `systemd` deliberately avoids importing `XDG_SESSION_ID` (I'm guessing there's a pretty sane reason for this though, given how that environment is shared across my own concurrent logins).

The bottom line is: If I use `systemd --user` to manage my user session, then my apps won't be a child of my login process, so `logind` won't work -- am I missing something here?

**Comment by keszybz** [comment]
systemctl / systemd have no notion of XDG_SESSION_ID, `git grep XDG_SESSION_ID` shows that only `pam_systemd` refers to it. So I think it must be some other reason why you don't have XDG_SESSION_ID imported.

But in general your reasoning is correct — if there are multiple sessions using the same XDG_SESSION_ID, they'd conflict over which one gets to set the variable, and there's no way for loginctl lock-session to know which session is meant, when it's running underneath some `systemd --user` service.

**Comment by WhyNotHugo** [comment]
Thanks!
